grudge/geometry/metrics.py

```diff
@@ -0,0 +1,710 @@
+
+
+def _signed_face_ones(
+        actx: ArrayContext, dcoll: DiscretizationCollection, dd
+        ) -> DOFArray:
+
+    assert dd.is_trace()
+
+    # NOTE: ignore quadrature_tags on dd, since we only care about
+    # the face_id here
+    all_faces_conn = dcoll.connection_from_dds(
+        DD_VOLUME, DOFDesc(dd.domain_tag)
+    )
+    signed_face_ones = dcoll.discr_from_dd(dd).zeros(
+        actx, dtype=dcoll.real_dtype
+    ) + 1
+    for igrp, grp in enumerate(all_faces_conn.groups):
+        for batch in grp.batches:
+            i = actx.thaw(batch.to_element_indices)
+            grp_field = signed_face_ones[igrp].reshape(-1)
+            grp_field[i] = \
+                (2.0 * (batch.to_element_face % 2) - 1.0) * grp_field[i]
+
+    return signed_face_ones
+
+
+def parametrization_derivative(
+        actx: ArrayContext, dcoll: DiscretizationCollection, dim, dd
+        ) -> MultiVector:
+    r"""Computes the product of forward metric derivatives spanning the
+    tangent space with topological dimension *dim*.
+
+    :arg dim: an integer denoting the spatial dimension. Defaults to the
+        :attr:`grudge.DiscretizationCollection.dim`.
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization.
+    :returns: a :class:`pymbolic.geometric_algebra.MultiVector` containing
+        the product of metric derivatives.
+    """
+    if dim == 0:
+        from pymbolic.geometric_algebra import get_euclidean_space
+
+        return MultiVector(
+            _signed_face_ones(actx, dcoll, dd),
+            space=get_euclidean_space(dcoll.ambient_dim)
+        )
+
+    from pytools import product
+
+    return product(
+        forward_metric_derivative_mv(actx, dcoll, rst_axis, dd)
+        for rst_axis in range(dim)
+    )
+
+
+def pseudoscalar(actx: ArrayContext, dcoll: DiscretizationCollection,
+        dim=None, dd=None) -> MultiVector:
+    r"""Computes the field of pseudoscalars for the domain/discretization
+    identified by *dd*
+
+    :arg dim: an integer denoting the spatial dimension. Defaults to the
+        :attr:`grudge.DiscretizationCollection.dim`.
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization.
+    :returns: A :class:`~pymbolic.geometric_algebra.MultiVector` of
+        :class:`~meshmode.dof_array.DOFArray`\ s
+    """
+    if dd is None:
+        dd = DD_VOLUME
+
+    if dim is None:
+        dim = dcoll.discr_from_dd(dd).dim
+
+    return parametrization_derivative(
+        actx, dcoll, dim, dd
+    ).project_max_grade()
+
+
+@memoize_on_first_arg
+def area_element(
+        actx: ArrayContext, dcoll: DiscretizationCollection, dim=None, dd=None
+        ) -> DOFArray:
+    r"""Computes the measure of a transformed element. These are used
+    to transform integrals from global to reference space, and are
+    commonly referred to as Jacobian determinants.
```

Comment:
I find the term "measure" here a bit odd, since it's technically [not one](https://en.wikipedia.org/wiki/Measure_(mathematics)). And the Jacobian determinant bit is only true AFAIK for the volume case (ambient == topological dim). I like the second bit here more: This is the scale factor needed to transform an integral from reference to global.

Reply:
https://github.com/inducer/grudge/pull/74/commits/3651ce2c3c6f742c5f9a3ac82886a911e58bb188